### ISSUE TITLE
CollectionView improvements

### ISSUE THREAD

**Issue by Dolfik1** [fsprojects/Fabulous#736]
I tried to find out the cause of issue #653 and found some problems in CollectionView implementation.

1. Wrong template selector algorithm:
https://github.com/fsprojects/Fabulous/blob/3b22759f23fe8727737f3b594071455e5ea4e33a/Fabulous.XamarinForms/src/Fabulous.XamarinForms.Core/CustomControls.fs#L102-L112

For example, we have two items:
1. `View.ContentView(View.Label(text = "bla bla"))`
2. `View.ContentView(View.Image(source= "bla.png"))`

The current algorithm will determine them as same template what is wrong actually.
To solve this problem we should use other different approach. In android RecyclerView this behavior detected with [getItemViewType](https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.ViewHolder#getItemViewType()) function and we can use same approach.

2. A collection item is created through an activator that does not firing `created` callback and does not set ViewRef value.

https://github.com/fsprojects/Fabulous/blob/3b22759f23fe8727737f3b594071455e5ea4e33a/Fabulous.XamarinForms/src/Fabulous.XamarinForms.Core/CustomControls.fs#L91-L97

I applied dirty hack, and solved issue #653 
```fsharp

type ViewElementDataTemplate(``type``, createFn: unit -> obj) =
    inherit DataTemplate(fun () ->
        let bindableObject = createFn () :?> BindableObject
        let onBindingContextChanged = BindableHelpers.createOnBindingContextChanged bindableObject
        bindableObject.BindingContextChanged.Add (fun _ -> onBindingContextChanged ())
        bindableObject :> obj
    )

type ViewElementDataTemplateSelector() =
    inherit DataTemplateSelector()
    let cache = Dictionary<Type, ViewElementDataTemplate>()
    override this.OnSelectTemplate(item, _) =
        let holder = item :?> IViewElementHolder
        let targetType = holder.ViewElement.TargetType
        match cache.TryGetValue targetType with
        | false, _ ->
            let template = ViewElementDataTemplate(targetType, fun _ -> holder.ViewElement.Create())
            cache.[targetType] <- template
            template :> DataTemplate
        | true, template ->
            template :> DataTemplate
```

But ofcourse this is very weak hack which will not work with recycling.

What do you think about improving CollectionView implementation?

**Comment by TimLariviere** [comment]
>Wrong template selector algorithm:
>For example, we have two items:
>
>    View.ContentView(View.Label(text = "bla bla"))
>    View.ContentView(View.Image(source= "bla.png"))
>
>The current algorithm will determine them as same template what is wrong actually.
>To solve this problem we should use other different approach. In android RecyclerView this behavior detected with getItemViewType function and we can use same approach.

`ViewElementDataTemplateSelector` is correct in itself. To understand why it's doing what's it's doing, you first need to understand how DataTemplate works in Xamarin.Forms.

ListView/CollectionView follow those steps:
- Set the `ItemTemplate` property to tell what a row should look like for a given item (which could directly be a DataTemplate or a DataTemplateSelector)
- Set the `ItemsSource` property with arbitrary data that will be available as a `BindingContext` inside the row
- Whenever a row needs to be displayed, ListView/CollectionView will request a DataTemplate from `ItemTemplate`
- Thanks to virtualization, Xamarin.Forms will try to reuse rows that got out of screen, or when the item changed. To reuse a row, `ItemTemplate` needs to give the exact same DataTemplate instance that was used to create the row, otherwise Xamarin.Forms will create a new row with the new DataTemplate.
- Once the row is created or reused, Xamarin.Forms will set the row's `BindingContext` property with the item from `ItemsSource`.

That's usually here the bindings can listen to `BindingContextChanged` and update with the new data accordingly.

NOTE: BindingContext is not exclusively used by bindings despite the name.

Now comes Fabulous.

When writing
```fs
View.CollectionView(
    items = [
        View.ContentView(View.Label("bla bla"))
        View.ContentView(View.Image("bla.png"))
    ]
)
```

Fabulous will:
- Instantiate the CollectionView
- Set `ItemTemplate = ViewElementDataTemplateSelector()`
- Set `ItemsSource = ObservableCollection<ViewElementHolder>` where `ViewElementHolder.ViewElement` is the control you declared (e.g. `View.ContentView(View.Label("bla bla"))`).

So when Xamarin.Forms wants to display a row for the first item (the label), it asks `ViewElementDataTemplateSelector` for a DataTemplate.

While we could directly return `DirectViewElementDataTemplate(holder.ViewElement)`, the thing with DataTemplate is that if the instance is not exactly the same, Xamarin.Forms will consider it can't reuse rows and will destroy the ones going out of screen and create new ones.
So this is basically destroying all performance optimizations of the row recycling.
Also Fabulous will never be able to differentially update the rows since Xamarin.Forms will destroy them whenever the items change.

To avoid that, Fabulous will only give Xamarin.Forms the thing it needs at minimum: an empty row of the correct type.
So `View.ContentView(View.Label("bla bla"))` and `View.ContentView(View.Image("bla.png"))` will have the exact same DataTemplate and Xamarin.Forms will be able to recycle the rows: a DataTemplate creating an empty `ContentView`.

To apply the content of this empty ContentView, Fabulous listens for the row's `BindingContext` since Xamarin.Forms will always set the item as the binding context.
Whenever the binding context changes, Fabulous will apply the differential update to either fill the empty ContentView or change the content to match the new requested one.

That's why `ViewElementDataTemplateSelector` only cares about the root type and not the actual ViewElement. Because the actual ViewElement will be taken into account when `BindingContextChanged` is raised.
But to be able to set the `BindingContext`, Xamarin.Forms first needs the root control instance, the empty ContentView.

For more information, please see https://github.com/fsprojects/Fabulous/issues/743#issuecomment-630211221

>A collection item is created through an activator that does not firing created callback and does not set ViewRef value.

For the reasons I explained above, you can't use the created callback inside `ViewElementDataTemplateSelector`.
Because each create callback is unique to one specific ViewElement, so you'll always need to create a new DataTemplate - disabling row recycling and differential updates.

>I applied dirty hack, and solved issue #653

That's interesting though.
`ViewRef` is indeed only set by `ViewElement.Create()` and since we can't do that in ListView/CollectionView, we'll need to expose the things `ViewElement.Create()` does so we can apply `ref` and `created` inside the rows.

**Comment by TimLariviere** [comment]
For the reasons explained above and in #743, the way ViewElementDataTemplateSelector is implemented shouldn't be an issue.
For the 2nd point, let's keep it in #653 and fix the problem there.
Closing the issue here.
